internal/fs/dir_handle.go

```diff
@@ -242,42 +248,53 @@ func (dh *dirHandle) ensureEntries(ctx context.Context) (err error) {
 //
 // LOCKS_REQUIRED(dh.Mu)
 // LOCKS_EXCLUDED(du.in)
-func (dh *dirHandle) ReadDir(
+func (dh *DirHandle) ReadDir(
 	ctx context.Context,
 	op *fuseops.ReadDirOp) (err error) {
 	// If the request is for offset zero, we assume that either this is the first
 	// call or rewinddir has been called. Reset state.
+
 	if op.Offset == 0 {
-		dh.entries = nil
-		dh.entriesValid = false
-	}
+		dh.Entries = nil
+		dh.EntriesValid = false
+		rec.Lock()
+		rec.length = 0
+		rec.err = nil
+		rec.entryForSorting = fuseutil.Dirent{}
+		rec.Unlock()
+		go dh.FetchEntriesAsync(fuseops.RootInodeID, true)
 
-	//when the offset is zero(new call made means fetch)
-	//when the offset is non zero and len(dh.entries ) - offset <0 means not enough data to serve the call
-	if !dh.entriesValid || (len(dh.entries) <= int(op.Offset) && op.Offset != 0 && ContinuationToken != "") {
-		err = dh.ensureEntries(ctx)
-		if err != nil {
+	}
+	rec.Lock()
+	if rec.length <= int(op.Offset) && (op.Offset == 0 || ContinuationToken != "") {
+		rec.cond.Wait()
```

Comment:
you are taking a lock and waiting for the condition to trigger. how will fetchEntriesAsync ever get the lock on rec?

Reply:
When we wait on the condition variable, the goroutine suspends . Internally cond.Wait() unlocks it and it locks it again only when it wakes up by other go routine through a signal/broadcast.

Reply:
can you add that comment.